strawberryfields/engine.py

```diff
@@ -531,161 +544,542 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
         return result
 
 
-class StarshipEngine(BaseEngine):
+class RequestFailedError(Exception):
+    """Raised when a request to the remote platform returns an error response."""
+
+
+class InvalidJobOperationError(Exception):
+    """Raised when an invalid operation is performed on a job."""
+
+
+class JobFailedError(Exception):
+    """Raised when a remote job enters a 'failed' status."""
+
+
+class JobStatus(enum.Enum):
+    """Represents the status of a remote job.
+
+    This class maps a set of job statuses to the string representations returned by the
+    remote platform.
     """
-    Starship quantum program executor engine.
 
-    Executes :class:`.Program` instances on the chosen remote backend, and makes
-    the results available via :class:`.Result`.
+    OPEN = "open"
+    QUEUED = "queued"
+    CANCELLED = "cancelled"
+    COMPLETE = "complete"
+    FAILED = "failed"
+
+    @property
+    def is_final(self) -> bool:
+        """Checks if this status represents a final and immutable state.
+
+        This method is primarily used to determine if an operation is valid for a given
+        status.
+
+        Returns:
+            bool: ``True`` if the job status is final, and ``False`` otherwise
+        """
+        return self in (JobStatus.CANCELLED, JobStatus.COMPLETE, JobStatus.FAILED)
+
+
+class Job:
+    """Represents a remote job that can be queried for its status or result.
+
+    This object should typically not be instantiated directly, but returned by an
+    `Engine` or `Connection` when a job is run.
 
     Args:
-        backend (str, BaseBackend): name of the backend, or a pre-constructed backend instance
-        polling_delay_seconds (int): the number of seconds to wait between queries when polling for
-            job results
+        id_ (str): the job ID
+        status (strawberryfields.engine.JobStatus): the job status
+        connection (strawberryfields.engine.Connection): the connection over which the
+            job is managed
     """
 
-    # This engine will execute jobs remotely.
-    REMOTE = True
+    def __init__(self, id_: str, status: JobStatus, connection: "Connection"):
+        self._id = id_
+        self._status = status
+        self._connection = connection
+        self._result = None
+
+    @property
+    def id(self) -> str:
+        """The job ID.
+
+        Returns:
+            str: the job ID
+        """
+        return self._id
+
+    @property
+    def status(self) -> JobStatus:
+        """The job status.
+
+        Returns:
+            strawberryfields.engine.JobStatus: the job status
+        """
+        return self._status
+
+    @property
+    def result(self) -> Result:
+        """The job result.
 
-    def __init__(self, backend, polling_delay_seconds=1, **kwargs):
-        super().__init__(backend)
+        This is only defined for complete jobs, and raises an exception for any other
```

Comment:
```suggestion
        This is only defined for completed jobs and raises an exception for any other
```

Reply:
Oh, this suggestion regards other parts as well, such as `JobStatus.COMPLETE`. I thought it was more descriptive to have `completed`, but this is pretty subjective and let me know if you'd rather keep it as is :)

Reply:
No, I totally agree that "completed" makes more sense in this context (as a synonym for "finished"), while "complete" sounds to me along the lines of "full" or "whole".

The reason I had it as "complete" to begin with is because that is the identifier used by the remote platform :smiley: So the value of the `JobStatus` enum would have to be `"complete"` to match up with the responses, but the enum name could certainly be `COMPLETED` (along with comments, docstrings, etc.). I will make the change, thanks!

Reply:
Oh, I see! Feel free to leave it as is, if you'd rather not have any further confusions coming from this difference :)

Reply:
I've updated all occurrences of "complete" to "completed" except for the single value in the enum member. Don't think it should lead to much confusion, but I'll add a comment to clarify,

Reply:
Great, thanks! :)